File: OneCodeTeam/Story-Creator-Sample-Application-for-Windows-14638/[C#][C++]-Story Creator Sample Application for Windows Phone 7 (CSWP7AzureVideoStory)/C# and C++/C#,C++/CSWP7AzureVideoStory/VideoStoryCreator/ViewModels/ComposePhotoViewModel.cs
/****************************** Module Header ******************************\
* Module Name:	ComposePhotoViewModel.cs
* Project:		VideoStoryCreator
* Copyright (c) Microsoft Corporation.
* 
* A view model used in the compose page.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Controls;
using VideoStoryCreator.Models;
using VideoStoryCreator.Transitions;

namespace VideoStoryCreator.ViewModels
{
    public class ComposePhotoViewModel : PhotoViewModel
    {
        public ITransition Transition { get; set; }

        private UserControl _transitionDesigner;
        public UserControl TransitionDesigner
        {
            get
            {
                return this._transitionDesigner;
            }
            set
            {
                if (this._transitionDesigner != value)
                {
                    this._transitionDesigner = value;
                    this.NotifyPropertyChange("TransitionDesigner");
                }
            }
        }

        public List<string> AvailableTransitions
        {
            get
            {
                return TransitionFactory.AvailableTransitions;
            }
        }

        /// <summary>
        /// Clone this view model.
        /// </summary>
        /// <returns>A copy of this view model.</returns>
        public ComposePhotoViewModel CopyTo()
        {
            ComposePhotoViewModel copy = new ComposePhotoViewModel()
            {
                Name = this.Name,
                PhotoDuration = this.PhotoDuration,
                TransitionDuration = this.TransitionDuration,
                MediaStream = this.MediaStream
            };
            if (this.Transition != null)
            {
                copy.Transition = this.Transition.Clone();
            }
            return copy;
        }

        /// <summary>
        /// Update values of this view model from a copy.
        /// </summary>
        public void CopyFrom(ComposePhotoViewModel source)
        {
            this.Name = source.Name;
            this.PhotoDuration = source.PhotoDuration;
            this.TransitionDuration = source.TransitionDuration;
            if (source.Transition != null)
            {
                this.Transition = source.Transition.Clone();
            }
            this.MediaStream = source.MediaStream;
        }

        /// <summary>
        /// Converts a model class to a view model class.
        /// </summary>
        public static ComposePhotoViewModel CreateFromModel(Photo model)
        {
            ComposePhotoViewModel viewModel = new ComposePhotoViewModel()
            {
                Name = model.Name,
                MediaStream = model.ThumbnailStream,
                PhotoDuration = (int)(model.PhotoDuration.TotalSeconds)                    
            };
            if (model.Transition != null)
            {
                viewModel.Transition = model.Transition;
                viewModel.TransitionDuration = (int)(model.Transition.TransitionDuration.TotalSeconds);
            }
            return viewModel;
        }

        /// <summary>
        /// Update the model corresponding to this view model.
        /// </summary>
        public void UpdateModel()
        {
            Photo photoToModify = App.MediaCollection.Where(p => p.Name == this.Name).FirstOrDefault();
            if (photoToModify != null && this.Transition != null)
            {
                photoToModify.Transition = this.Transition;
                photoToModify.Transition.TransitionDuration = TimeSpan.FromSeconds(this.TransitionDuration);
                photoToModify.PhotoDuration = TimeSpan.FromSeconds(this.PhotoDuration);
            }
        }

        /// <summary>
        ///  If the view model is removed, we need to remove the corresponding model as well.
        /// </summary>
        public void RemoveModel()
        {
            Photo model = App.MediaCollection.Where(p => p.Name == this.Name).FirstOrDefault();
            if (model != null)
            {
                model.ThumbnailStream.Close();
                App.MediaCollection.Remove(model);
            }
        }
    }
}

File: OneCodeTeam/Story-Creator-Sample-Application-for-Windows-14992/[VB][C++]-Story Creator Sample Application for Windows Phone 7 (VBWPAzureVideoStory)/C++ and VB/C++,VB/VBWPAzureVideoStory/VideoStoryCreator/ViewModels/ComposePhotoViewModel.vb
'********************************* Module Header *********************************\
' Module Name: ComposePhotoViewModel.vb
' Project: VideoStoryCreator
' Copyright (c) Microsoft Corporation.
' 
' A view model used in the compose page.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'***************************************************************************/



Public Class ComposePhotoViewModel
    Inherits PhotoViewModel
    Public Property Transition() As ITransition
        Get
            Return m_Transition
        End Get
        Set(value As ITransition)
            m_Transition = value
        End Set
    End Property
    Private m_Transition As ITransition

    Private _transitionDesigner As UserControl
    Public Property TransitionDesigner() As UserControl
        Get
            Return Me._transitionDesigner
        End Get
        Set(value As UserControl)
            If Me._transitionDesigner IsNot value Then
                Me._transitionDesigner = value
                Me.NotifyPropertyChange("TransitionDesigner")
            End If
        End Set
    End Property

    Public ReadOnly Property AvailableTransitions() As List(Of String)
        Get
            Return TransitionFactory.AvailableTransitions
        End Get
    End Property

    ''' <summary>
    ''' Clone this view model.
    ''' </summary>
    ''' <returns>A copy of this view model.</returns>
    Public Function CopyTo() As ComposePhotoViewModel
        Dim copy As New ComposePhotoViewModel() With { _
          .Name = Me.Name, _
          .PhotoDuration = Me.PhotoDuration, _
          .TransitionDuration = Me.TransitionDuration, _
          .MediaStream = Me.MediaStream _
        }
        If Me.Transition IsNot Nothing Then
            copy.Transition = Me.Transition.Clone()
        End If
        Return copy
    End Function

    ''' <summary>
    ''' Update values of this view model from a copy.
    ''' </summary>
    Public Sub CopyFrom(source As ComposePhotoViewModel)
        Me.Name = source.Name
        Me.PhotoDuration = source.PhotoDuration
        Me.TransitionDuration = source.TransitionDuration
        If source.Transition IsNot Nothing Then
            Me.Transition = source.Transition.Clone()
        End If
        Me.MediaStream = source.MediaStream
    End Sub

    ''' <summary>
    ''' Converts a model class to a view model class.
    ''' </summary>
    Public Shared Function CreateFromModel(model As Photo) As ComposePhotoViewModel
        Dim viewModel As New ComposePhotoViewModel() With { _
          .Name = model.Name, _
          .MediaStream = model.ThumbnailStream, _
          .PhotoDuration = CInt(model.PhotoDuration.TotalSeconds) _
        }
        If model.Transition IsNot Nothing Then
            viewModel.Transition = model.Transition
            viewModel.TransitionDuration = CInt(model.Transition.TransitionDuration.TotalSeconds)
        End If
        Return viewModel
    End Function

    ''' <summary>
    ''' Update the model corresponding to this view model.
    ''' </summary>
    Public Sub UpdateModel()
        Dim photoToModify As Photo = App.MediaCollection.Where(Function(p) p.Name = Me.Name).FirstOrDefault()
        If photoToModify IsNot Nothing AndAlso Me.Transition IsNot Nothing Then
            photoToModify.Transition = Me.Transition
            photoToModify.Transition.TransitionDuration = TimeSpan.FromSeconds(Me.TransitionDuration)
            photoToModify.PhotoDuration = TimeSpan.FromSeconds(Me.PhotoDuration)
        End If
    End Sub

    ''' <summary>
    '''  If the view model is removed, we need to remove the corresponding model as well.
    ''' </summary>
    Public Sub RemoveModel()
        Dim model As Photo = App.MediaCollection.Where(Function(p) p.Name = Me.Name).FirstOrDefault()
        If model IsNot Nothing Then
            model.ThumbnailStream.Close()
            App.MediaCollection.Remove(model)
        End If
    End Sub
End Class
